File: Fundamentals_of_Software_Design/Python_language/Rabin_Karp/rabin_karp.py
def polynomial_hash(s, base=256, modulus=101):
    hash_value = 0
    for char in s:
        hash_value = (hash_value * base + ord(char)) % modulus
    return hash_value

def update_hash(prev_hash, prev_char, new_char, length, base=256, modulus=101):
    prev_hash = (prev_hash - ord(prev_char) * pow(base, length-1, modulus)) % modulus
    return (prev_hash * base + ord(new_char)) % modulus

def my_strstr(haystack, needle):
    if not needle:
        return 0
    if not haystack or len(needle) > len(haystack):
        return -1
    
    n = len(needle)
    m = len(haystack)
    base = 256
    modulus = 101
    
    needle_hash = polynomial_hash(needle, base, modulus)
    window_hash = polynomial_hash(haystack[:n], base, modulus)
    
    if needle_hash == window_hash and haystack[:n] == needle:
        return 0
    
    for i in range(1, m - n + 1):
        window_hash = update_hash(
            window_hash,
            haystack[i-1],
            haystack[i+n-1],
            n,
            base,
            modulus
        )
        
        if needle_hash == window_hash and haystack[i:i+n] == needle:
            return i
    
    return -1

File: Fundamentals_of_Software_Design/C_language/Rabin_Karp/rabin_karp.c
#include <stdlib.h>
#include <string.h>

#include "rabin_karp.h"

#define BASE 256
#define MODULUS 101

static unsigned int hash(const char* str, size_t len) {
    unsigned int h = 0;
    for (size_t i = 0; i < len; i++) {
        h = (h * BASE + str[i]) % MODULUS;
    }
    return h;
}

static unsigned int update_hash(unsigned int prev_hash, char prev_char, char new_char, size_t len) {
    unsigned int h = (prev_hash + MODULUS - (prev_char * (unsigned int)pow(BASE, len-1) % MODULUS) % MODULUS;
    return (h * BASE + new_char) % MODULUS;
}

const char* my_strstr(const char* haystack, const char* needle) {
    if (!haystack || !needle) return NULL;
    
    size_t n = strlen(haystack);
    size_t m = strlen(needle);
    
    if (m == 0) return haystack;
    if (n < m) return NULL;
    
    unsigned int needle_hash = hash(needle, m);
    unsigned int window_hash = hash(haystack, m);
    
    for (size_t i = 0; i <= n - m; i++) {
        if (window_hash == needle_hash && 
            memcmp(haystack + i, needle, m) == 0) {
            return haystack + i;
        }
        
        if (i < n - m) {
            window_hash = update_hash(window_hash, 
                                    haystack[i], 
                                    haystack[i + m], 
                                    m);
        }
    }
    
    return NULL;
}
